pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+
+	var files Files
+	for i, m := range maps {
+		if i > 0 && m.Pathname == maps[i-1].Pathname {
+			// skip duplicate entries
+			continue
+		}
+		if m.Pathname == "" {
+			// anonymous mapping
+			continue
+		}
+
+		if m.Dev == 0 || m.Inode == 0 {
+			// virtual memory region, eg. [heap], [stack], [vvar], etc
+			continue
+		}
+
+		file := File{
+			Name:     m.Pathname,
+			OpenPerm: permToString(m.Perms),
+		}
+
+		if file.Type, file.FilePerm, file.Size, _ = fileStats(ofl.stat, m.Pathname); file.Type == "" {
+			continue
+		}
+		file.Fd = mmapFD(m.Pathname, file.Type, cwd)
+
+		files = append(files, file)
+	}
+
+	return files, nil
+}
+
+func permToString(perms *procfs.ProcMapPermissions) string {
+	s := ""
+
+	for _, perm := range []struct {
+		set       bool
+		charSet   string
+		charUnset string
+	}{
+		{perms.Read, "r", "-"},
+		{perms.Write, "w", "-"},
+		{perms.Execute, "x", "-"},
+		{perms.Private, "p", ""},
+		{perms.Shared, "s", ""},
+	} {
+		if perm.set {
+			s += perm.charSet
+		} else {
+			s += perm.charUnset
+		}
+	}
+
+	return s
+}
+
+func mmapFD(path string, ty, cwd string) string {
+	/*
+		cwd  current working directory;
+		ltx  shared library text (code and data);
+		mem  memory-mapped file;
+		mmap memory-mapped device;
+		rtd  root directory;
+		txt  program text (code and data);
+	*/
+	fd := "unknown"
+	if ty == "REG" {
+		// knowing whether the file is memory mapped or program text would require knowing permissions
+		// which aren't parsed by gopsutil, so we just assume it's memory mapped
+		fd = "mem"
+	} else if ty == "DIR" {
+		if path == "/" {
+			fd = "rtd"
+		} else if cwd != "" && path == cwd {
+			fd = "cwd"
+		}
+	}
+	return fd
+}
+
+func (ofl *openFilesLister) fdMetadata() (Files, error) {
+	openFDs, err := ofl.proc.FileDescriptors()
+	if err != nil {
+		return nil, err
+	}
+
+	var files Files
+	for _, openFD := range openFDs {
+		file, ok := ofl.fdStat(openFD)
+		if ok {
+			files = append(files, file)
+		}
+	}
+
+	return files, nil
+}
+
+func (ofl *openFilesLister) fdStat(fd uintptr) (File, bool) {
+	var err error
+	file := File{
+		Fd: fmt.Sprintf("%d", fd),
+	}
+
+	fdLinkPath := fmt.Sprintf("%s/fd/%d", ofl.procPIDPath(), fd)
+
+	var inode uint64
+	if file.Type, file.OpenPerm, _, _ = fileStats(ofl.lstat, fdLinkPath); file.Type == "" {
+		return File{}, false
+	}
+
+	var ok bool
+	// remove some unnecessary information from permissions string
+	// if the expectations are not met, the permissions are left as is
+
+	// file descriptors always have no sticky bit, setuid, setgid
+	file.OpenPerm = strings.TrimPrefix(file.OpenPerm, "-")
+	// file descriptors always have no permission for group and others
+	file.OpenPerm, ok = strings.CutSuffix(file.OpenPerm, "------")
+	if ok {
+		// file descriptors always have execute permission
+		file.OpenPerm = strings.TrimSuffix(file.OpenPerm, "x")
+	}
+
+	if file.Type, file.FilePerm, file.Size, inode = fileStats(ofl.stat, fdLinkPath); file.Type == "" {
```

Comment:
Should we return a `fileStat` struct from the function `fileStats` rather than four unstructured values?

Reply:
Yep fair

Reply:
Actually that makes the code bigger overall as I can't assign return values into existing variables anymore, ie. this
```go
if file.Type, file.FilePerm, file.Size, _ = fileStats(ofl.stat, m.Pathname); file.Type == "" {
	continue
}
```
has to become something like
```go
stats := fileStats(ofl.stat, m.Pathname)
if stats.Type == "" {
	continue
}
file.Type = stats.Type
file.FilePerm = stats.FilePerm
file.Size = stats.Size
```

What do you think ? Does it seem reasonable to keep it the current way ?

Reply:
Yeah, absolutely. In the end it boils down to personal preference, and how often do we plan on updating this code. I would imagine that we are not going to touch it very often.